pkg/graveler/graveler.go

```diff
@@ -1487,17 +1499,36 @@ func (g *Graveler) DeleteBatch(ctx context.Context, repository *RepositoryRecord
 	if len(keys) > DeleteKeysMaxSize {
 		return fmt.Errorf("keys length (%d) passed the maximum allowed(%d): %w", len(keys), DeleteKeysMaxSize, ErrInvalidValue)
 	}
-	var m *multierror.Error
+
 	err = g.safeBranchWrite(ctx, g.log(ctx).WithField("operation", "delete_keys"),
 		repository, branchID, func(branch *Branch) error {
-			var cachedMetaRangeID MetaRangeID // used to cache the committed branch metarange ID
+			commit, err := g.RefManager.GetCommit(ctx, repository, branch.CommitID)
+			if err != nil {
+				return err
+			}
+			cachedMetaRangeID := commit.MetaRangeID
+
+			deleteWorkers := int(math.Min(float64(100), float64(len(keys))))
+			keysChan := make(chan Key, len(keys))
+			var wg multierror.Group
+			cctx, cancel := context.WithCancel(ctx)
+			defer cancel()
+			for i := 0; i < deleteWorkers; i++ {
+				wg.Go(func() error {
+					return g.deleteConsumer(cctx, cancel, repository, branch, &cachedMetaRangeID, keysChan)
+				})
+			}
+
 			for _, key := range keys {
-				err := g.deleteUnsafe(ctx, repository, branch, key, &cachedMetaRangeID)
-				if err != nil {
-					m = multierror.Append(m, &DeleteError{Key: key, Err: err})
+				select {
+				case <-cctx.Done():
+					break
```

Comment:
This will cause issues if you ever cancel: you never abort any of the planned workers.  So every delete worker will run until keysChan is empty.

We could equally well write
```go
			for _, key := range keys {
				keysChan <- key
			}
```
They are the same effectively because realistically ctx will not be cancelled before every key is loaded on keysChan -- blocking is impossible because keysChan has capacity for all of keys.

I think handling cancellation _inside_ the workers on line 1518 is the important part, and the only bit that will run in real cases.